cachedcredentials/cachedcredentials.go

```diff
@@ -0,0 +1,162 @@
+func fileExists(name string) (bool, error) {
+	_, err := os.Stat(name)
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+	return true, err
+}
+
+// writeCacheFile writes cachedCredentials to a file atomically
+func writeCacheFile(cacheFilename string, cachedCreds cachedCredentials) error {
+	data, err := json.Marshal(cachedCreds)
+	if err != nil {
+		return err
+	}
+
+	fd, err := syscall.Open(cacheFilename, syscall.O_CREAT|syscall.O_RDWR, 0600)
+	defer syscall.Close(fd)
+	if err != nil {
+		message := fmt.Sprintf("Encountered error while opening credentials file %s\n", cacheFilename)
+		os.Stderr.WriteString(message)
+		os.Stderr.WriteString(err.Error() + "\n")
+		return err
+	}
+
+	err = syscall.Flock(fd, syscall.LOCK_EX)
+	if err != nil {
+		message := fmt.Sprintf("Encountered error while requesting a lock on credentials file %s\n", cacheFilename)
+		os.Stderr.WriteString(message)
+		os.Stderr.WriteString(err.Error() + "\n")
+		return err
+	}
+
+	_, writeErr := syscall.Write(fd, data)
+
+	err = syscall.Flock(fd, syscall.LOCK_UN)
+
+	if writeErr != nil {
+		message := fmt.Sprintf("Encountered error while writing credentials file %s\n", cacheFilename)
+		os.Stderr.WriteString(message)
+		os.Stderr.WriteString(err.Error() + "\n")
+		return writeErr
+	}
+
+	if err != nil {
+		message := fmt.Sprintf("Encountered error while unlocking credentials file %s\n", cacheFilename)
+		os.Stderr.WriteString(message)
+		os.Stderr.WriteString(err.Error() + "\n")
+		return err
+	}
+
+	return nil
+}
+
+func readCacheFile(cacheFilename string) (cachedCredentials, error) {
+	cachedCreds := cachedCredentials{}
+	bytes, err := ioutil.ReadFile(cacheFilename)
+	if err != nil {
+		message := fmt.Sprintf("Encountered error while reading cached credentials file %s\n", cacheFilename)
+		os.Stderr.WriteString(message)
+		os.Stderr.WriteString(err.Error() + "\n")
+		return cachedCreds, err
+	}
+	err = json.Unmarshal(bytes, &cachedCreds)
+	if err != nil {
+		message := fmt.Sprintf("Encountered error while parsing cached credentials file %s\n", cacheFilename)
+		os.Stderr.WriteString(message)
+		os.Stderr.WriteString(err.Error() + "\n")
+		return cachedCreds, err
+	}
+	return cachedCreds, err
+}
+
+func (f *FileCacheProvider) refreshCredentials(cacheFilename string) (cachedCredentials, error) {
+	credentials, err := f.Creds.Get()
+	if err != nil {
+		return cachedCredentials{}, err
+	}
+
+	expiresAt, err := f.Creds.ExpiresAt()
+	if err != nil {
+		return cachedCredentials{}, err
+	}
+
+	cachedCreds := cachedCredentials{
+		AccessKeyID:     credentials.AccessKeyID,
+		ExpiresAt:       expiresAt,
+		ProviderName:    credentials.ProviderName,
+		SecretAccessKey: credentials.SecretAccessKey,
+		SessionToken:    credentials.SessionToken,
+	}
+
+	err = writeCacheFile(cacheFilename, cachedCreds)
+	return cachedCreds, err
+}
+
+// Retrieve retrieves credentials
+func (f *FileCacheProvider) Retrieve() (credentials.Value, error) {
+	cacheDir := path.Join(os.Getenv("HOME"), ".s3parcp")
```

Comment:
consider using `$XDG_CACHE_HOME/.s3parcp` which would default to `~/.cache/.s3parcp`

Reply:
Also, HOME is not guaranteed to be set. The way to get the user home directory is `os.UserHomeDir()` apparently: https://stackoverflow.com/a/55491425/1380386